### ISSUE TITLE
[Feat] Add result (simulate, filtering, training process) saving functionality

### ISSUE THREAD

**Issue by hanson-hschang** [hanson-hschang/Signal-System#65]
## Observer pattern

Observer (a.k.a. publish/subscribe or listener pattern) decouples a stateful object (the **subject**/**observable**) from the code that reacts to its changes (the **observers**). The subject holds a list of observers and exposes `attach()`/`detach()`; whenever something noteworthy happens it calls `notify()`, which invokes an `update()` (or similarly-named) hook on every registered observer. The subject never needs to know what its observers actually do — logging, plotting, persistence, etc. — which is the main win: you can add a new observer without touching the subject's code (open/closed principle), and swap observers in and out at runtime.

The two-line version: subject changes state → subject shouts "something happened" → observers each decide what to do about it.

## Is it a good fit here? — Short answer: not for `simulate`/`filtering`, but yes for your training loop

I pulled apart `ss-jax` to check, and the honest answer depends on *where* in the codebase you'd apply it.

**Your old PyTorch/Numba loop was already observer-shaped.** Look at your `__main__.py`:

```python
for k in tqdm(range(simulation_time_steps)):
    system_callback.record(k, current_time)
    current_time = markov_chain.process(current_time)
```

This is an eager Python `for` loop, and `callback.record()` is effectively a single hard-coded observer being notified every iteration. `Callback` conflates "the observer" and "the thing being observed" into one imperative loop — which works fine because the loop runs step-by-step in real Python.

**`ss-jax`'s `simulate()` and `filtering()` are not eager loops — they're `jax.lax.scan` under `@jax.jit`.** I checked `_system.py` and `_filtering.py`: the entire rollout happens inside a traced, compiled `scan` body, and the function returns the *whole stacked trajectory* (`times, states, observations, controls`) in one shot. There is no per-step Python-level hook you can attach an observer to without breaking things:

- You can't call arbitrary Python object methods (like `observer.update(...)`) from inside a `scan` body — it isn't traced, so it either errors or silently only runs once during tracing.
- `jax.debug.callback` exists for exactly this kind of host-side side-effect, but it round-trips to Python every step, which defeats the purpose of compiling the rollout into one XLA program. It's meant for occasional debug prints, not bulk per-step data collection.
- The only way to get genuine per-step Python control flow back would be to un-jit the loop — throwing away the main benefit of the migration you just did.

The deeper point: **`scan`'s second return value already *is* the recorder.** In the old code, the callback's job was to accumulate per-step values into a list because nothing else was doing that. In JAX, accumulation is what `scan` does natively — you already get `states`, `observations`, `beliefs`, `controls` fully materialized as arrays when `simulate`/`filtering` returns. There's nothing left for a per-step observer to *do*. Even `step_skip` subsampling, which the old `Callback.record()` handled with a per-step `if current_step % sample_every == 0`, becomes a trivial post-hoc slice (`states[::step_skip]`) once you have the full array — no need for conditional logic that JAX would force through `lax.cond` anyway.

So: implementing Observer around the simulate/filter rollout would be fighting the architecture, not adopting a pattern. What you actually want there is much simpler than what you're picturing.

**Where Observer genuinely earns its keep is `LearningProcess.train_one_epoch`** in `utility/learning/_process.py`. That *is* a real eager Python `for batch in batch_loader` loop — no `jit`/`scan` wrapping the whole thing, just per-step `train_step` calls. Right now it hardcodes bookkeeping (`self._info.training_losses.append(...)`, a tqdm postfix). That's precisely the shape Observer is meant for, and it lines up with what I remember you were already building — MLflow/orbax/TensorBoard tracking — as separate concerns bolted onto one loop.

## Recommendation

Two different patterns for two different parts of the codebase:

**1. Simulation/filtering results → HDF5: a small `TrajectoryData` result object, not Observer.**
Since `simulate`/`filtering` already hand you complete arrays, collapse your old `Callback` + `Data` split into one lightweight `eqx.Module`-adjacent container built directly from those outputs:

```python
@dataclass
class TrajectoryData:
    signal_trajectory: dict[str, Array]   # times, states, observations, beliefs, controls...
    meta_data: MetaData = field(default_factory=MetaData)
    meta_info: MetaInfo = field(default_factory=MetaInfo)

    @classmethod
    def from_simulation(cls, times, states, observations, controls=None, **meta) -> "TrajectoryData":
        traj = {"time": times, "state": states, "observation": observations}
        if controls is not None:
            traj["control"] = controls
        return cls(traj, ...)

    def save(self, filename: str | Path) -> None: ...   # reuse your existing HDF5 read/write logic
    @classmethod
    def load(cls, filename: str | Path) -> "TrajectoryData": ...
```

Your existing `_save_meta_data`/`_save_meta_info`/`_save_callback_params` static methods in `callback.py` port over almost unchanged — the hierarchical-key HDF5 group logic doesn't care whether the array came from a Python list or a JAX array. You just drop the `record()`/step-loop machinery entirely, since it has no job anymore.

If you want *some* extensibility here without full Observer, use **Strategy**, not Observer: let `save()` delegate to a pluggable `Writer` (`HDF5Writer`, later maybe `NpzWriter`), so you can swap formats without an ongoing "notify on every event" relationship — there's only ever one event (save), so Observer's multi-subscriber machinery is overkill.

**2. Training loop → real Observer pattern.**

```python
class TrainingObserver(Protocol):
    def on_epoch_end(self, epoch: int, info: LearningProcessInfo) -> None: ...
    def on_batch_end(self, iteration: int, loss: float) -> None: ...

class LearningProcess(Generic[Model]):
    def __init__(self, ..., observers: list[TrainingObserver] | None = None):
        self._observers = observers or []

    def _notify_batch_end(self, iteration, loss):
        for obs in self._observers:
            obs.on_batch_end(iteration, loss)
```

Then `MlflowObserver`, `TensorBoardObserver`, `OrbaxCheckpointObserver`, `ConsoleObserver` each implement the same interface independently — this is where decoupling actually pays off, since you genuinely want several independent side effects firing off one eager loop.

**Comment by hanson-hschang** [comment]
@skim0119 what do you think?

**Comment by skim0119** [comment]
Can we use one of the old example script that uses callback (from `main` branch), and prototype the user API with new syntax?

**Comment by hanson-hschang** [comment]
```python
transition = jnp.array([[0.7, 0.3], [0.4, 0.6]])
emission = jnp.array([[0.8, 0.2], [0.2, 0.8]])
system = HiddenMarkovModel(
    transition=ProbabilityParameter(transition),
    emission=ProbabilityParameter(emission),
)

random_key = jax.random.PRNGKey(0)
initial_state = system.initial_state(random_key)
times, states, observations, _ = simulate(system, 0, time_horizon, initial_state, random_key)

trajectory_data = SequenceData()

# meta info saves non-numerics meta information
trajectory_data.meta_info["system"] = "hidden markov model"

# meta data saves numberics meta information
trajectory_data.meta_data["transition_matrix"] = system.transition_matrix
trajectory_data.meta_data["emission_matrix"] = system.emission_matrix

# dictionary-like assignment
trajectory_data["times"] = times
trajectory_data["states"] = states
tajectory_data["observations"] =  observations

# save method api
trajectory_data.save("system.hdf5")
```

---

```python
# load from file
trajectory_data = SequenceData.load("system.hdf5")

# meta_info and meta_data access
print(trjaectory_data.meta_info["system"])
print(trajectory_data.meta_data["transition_matrix"].shape)
print(trajectory_data.meta_data["emission_matrix"].shape)

# dictionary-like access
print(trajectory_data["times"].shape)
print(trajectory_data["states"].shape)
print(trajectory_data["observations"].shape)
```

---

If there are "/" in a key, then it should be saved as a group in hdf5.

**Comment by hanson-hschang** [comment]
I already have the following `LearningProcessInfo` and this is where Im not sure how to integrate the observer pattern to incooporate different observers like `MlflowObserver`, `TensorBoardObserver`, `OrbaxCheckpointObserver`, `ConsoleObserver`, etc. 

```python

@dataclass
class LearningProcessInfo:
    training_losses: list[float] = field(default_factory=list)
    validation_losses: list[float] = field(default_factory=list)
    epoch: int = 0
    iteration: int = 0


class LearningProcess(Generic[Model]):
    def __init__(
        self,
        model: Model,
        loss_function: LossFunctionProtocol[Model],
        optimizer: optax.GradientTransformation,
        trainable_mask: PyTree | None = None,
    ) -> None:
        self.learning_state = LearningState[Model].create(model, optimizer, trainable_mask)
        self._train_step = make_train_step(loss_function)
        self._evaluate_step = make_evaluate_step(loss_function)
        self._info = LearningProcessInfo()

    @property
    def model(self) -> Model:
        return self.learning_state.model

    def train_one_epoch(
        self,
        training_data_loader: Iterable[Any],
        validation_data_loader: Iterable[Any] | None = None,
        random_key: PRNGKeyArray | None = None,
        validate_every: int | None = None,
        progress_bar: bool = True,
    ) -> None:
        batch_loader = (
            tqdm(training_data_loader, desc=f"epoch {self._info.epoch}") if progress_bar else training_data_loader
        )

        for batch in batch_loader:
            training_random_key, random_key = split_random_key(random_key)

            self.learning_state, loss = self._train_step(self.learning_state, batch, training_random_key)
            loss_value = float(loss)
            self._info.training_losses.append(loss_value)
            self._info.iteration += 1

            if progress_bar:
                batch_loader.set_postfix(loss=loss_value)  # type: ignore[union-attr]

            if (
                validation_data_loader is not None
                and validate_every is not None
                and self._info.iteration % validate_every == 0
            ):
                validation_random_key, random_key = split_random_key(random_key)
                validation_losses = self.evaluate_model(validation_data_loader, validation_random_key)
                self._info.validation_losses.append(float(jnp.mean(validation_losses)))

        self._info.epoch += 1
```

**Comment by hanson-hschang** [comment]
The user api for learning in the previous setting is as follows and the checkpoint files (learning process info) are saved automatically throughout the training session.

```python
learning_process = LearningProcess(...)

# Prepare training configuration
training_config = TrainingConfig()
training_config.validation.per_iteration_period = 300
training_config.validation.max_batch = 50
training_config.termination.max_epoch = 16

# the following is the setup for checkpoint saving
training_config.checkpoint.folderpath = "checkpoints/"
training_config.checkpoint.filename = "model"
training_config.checkpoint.per_epoch_period = 4

# Train model
learning_process.training(
    training_loader, evaluation_loader, training_config
)
```

**Comment by skim0119** [comment]
I think, data callback should be passed through 'simulate' or 'filter' directly, and recording should be managed internally. Meta data should be recorded without user-interrupt. Internally, it can have checkpoint parameters (file path, save interval) that 'simulate' or 'filter' use to record the state.

Probably simplest version would be just let 'simulate' and 'filter' to return dictionary of data, and add save/load function to I/O h5 files.

**Comment by hanson-hschang** [comment]
@skim0119 I think it might also be important to first get this carry class done? #66 Then, it will be easier to implement a record method in the class to save it to the file?

**Comment by hanson-hschang** [comment]
> Probably simplest version would be just let 'simulate' and 'filter' to return dictionary of data, and add save/load function to I/O h5 files.

Yeah, this is the simplest. It will be like following:

```python
times, states, observations, _ = simulate(system, 0, time_horizon, initial_state, random_key)

trajectory_data = SequenceData()

# dictionary-like assignment
trajectory_data["times"] = times
trajectory_data["states"] = states
tajectory_data["observations"] =  observations

# save method api
trajectory_data.save("system.hdf5")
```

However, it won't be the same as what you initially responded, right?
> data callback should be passed through 'simulate' or 'filter' directly, and recording should be managed internally.

**Comment by hanson-hschang** [comment]
> Meta data should be recorded without user-interrupt. Internally, it can have checkpoint parameters (file path, save interval) that 'simulate' or 'filter' use to record the state.

I agree, and it should be saved as something not easilly overwritten. In the codebase `main` branch, these are save with dunder names like `__transition_matrix__`, `__emission_matrix__`.
And not just in the `simulate` and `filter`, but for the training loop as well. The checkpoint parameters should be autommatically saved. Something like [this `training_config`](https://github.com/hanson-hschang/Signal-System/issues/65#issuecomment-5133882942) should be a good setting.

https://github.com/hanson-hschang/Signal-System/blob/c5b736d8f4504379ad4dd405cb1fd0f5fe7661ce/src/ss/utility/learning/process/config/_config.py#L173-L178

In `CheckpointConfig`:

https://github.com/hanson-hschang/Signal-System/blob/c5b736d8f4504379ad4dd405cb1fd0f5fe7661ce/src/ss/utility/learning/process/checkpoint/config/_config.py#L118-L131

**Comment by skim0119** [comment]
Question on learning process: 
How many variation of `LearningProcess` are you going to have? Can we just simplify to single `train` function like `simulate` and `filter`?

Thoughts on I/O
How about `np.savez` style:
```py
storage = SequenceData(H5_PATH)
storage.save(times=times, states=states, observations=observations, group="/simulate")
print(storage)

>> Storage
>> ├── Groups
>> │   └── /simulate
>> │       ├── times
>> │       ├── states
>> │       └── observations
>> └── Summary
>>     Groups:   1
>>     Arrays:   3
>>     Size:     42.1 KB
...

...

# restore
states[:] = storage["/simulate/states"]
```

`save`:
```py
    async def save(self, *, timestep=0, group="/", **kwargs):
        with h5py.File(self.fname, 'w') as f:
            for name, data in kwargs.items():
                # assert data and dtype are compatible
                group = f.create_group(group + f"/{timestep:04d}")
                group.create_dataset(name, data=data)
```

This could be inserted to `simulate` as
```py
simulate(system, 0, time_horizon, initial_state, random_key, db=storage, save_every=10)
```

**Comment by hanson-hschang** [comment]
1. For the learning process, Im not sure what you mean by "variation" of `LearningProcess`.
2. The I/O and the structure look good. A quick question, what happen if user want to save other information? The current setting looks like `storage.save(times=times, states=states, observations=observations, group="/simulate")` is hard coded inside the simulate function? If user want to save `control`, `control_state` or some other information, how to customize it?

**Comment by skim0119** [comment]
> For the learning process, Im not sure what you mean by "variation" of LearningProcess.

I'm just thinking, if it is necessary to wrap the learning process into the class. The fact that so many other modules needs to be passed to the parameter `training` might mean that it is doing too many things in one shot. I need to think more about the structure, I'll look at the previous example cases.

2. This is some thoughts for this prototype: in `simulate`, db should be passed to modules (`system` and `controller` etc.), and saving occurs within those modules when necessary. The amount of diagnostic recording can be controlled when we define those modules.
For instance,
```py
system = HiddenMarkovModel(
    transition=ProbabilityParameter(transition),
    emission=ProbabilityParameter(emission),
    diagnostic_level=2, <--
)
```
The reason is that giving full injection of function to customize logging is not very friendly with jax-looping. And most of the time, what people wants to record is somewhat fixed. Giving some pre-defined level of diagnostic capability might be better, and predictable (favor for both developer and jax). Just suggesting alternative direction.

**Comment by hanson-hschang** [comment]
For the learning process, here are my thoughts:

It requires two main functionality and two additional features:
- main functionalities:
    - `validate_model`: requires `model` and `loss_function`
    - `train_model`: requires `model`, `loss_function`, `optimizer`, and optionally `step_size_scheduler`
- additional features:
    -  `save_model`: save to checkpoint with info including model states, optimizer (and scheduler) states, training and validation loss history (over iterations)
    -  `load_model`: load from checkpoint with above info

Some related notes:
- The current learning process class have four init arguements: `model`, `loss_function`, `optimizer`, and optionally `step_size_scheduler`because it is required for training and save model. 
- For the functions `validate_model` and `train_model`, they should share one common thing: compute the loss for one batch. The current simplified implementation on the main branch (in pytorch) is like following:
```python
    # This should share with both training and validation
    def _evaluate_one_batch(
        self, data_batch: tuple[torch.Tensor, ...]
    ) -> torch.Tensor:
         loss = ...
         return loss

    def validate_model(
        self,
        data_loader: DataLoader[tuple[torch.Tensor, ...]],
    ) -> NDArray:
        losses = []
        for b, data_batch in enumerate(data_loader):
            evaluation_loss = self._evaluate_one_batch(data_batch).item()
            losses.append(evaluation_loss)

        return np.array(losses)

    def _train_one_batch(
        self, data_batch: tuple[torch.Tensor, ...]
    ) -> torch.Tensor:
        self._optimizer.zero_grad()
        loss_tensor = self._evaluate_one_batch(
            self._device_manager.load_data_batch(data_batch)
        )
        loss_tensor.backward()
        self._optimizer.step()
        return loss_tensor

    def train_model(
        self,
        training_data_loader: DataLoader[tuple[torch.Tensor, ...]],
        validation_data_loader: DataLoader[tuple[torch.Tensor, ...]],
    ) -> None:    
        for data_batch in training_data_loader:
            training_loss = self._train_one_batch(data_batch).item()
            validation_losses = self.validate_model(validation_data_loader)
```

---

There is more detail info required in the training and validation functions. For example:
- In training: how frequently to do validation and save to checkpoint? What are the termination conditions?
- In validation: how many batches required for each validation? It does not need to go through the whole validation dataset every time. 
These are provided in the `TrainingConfig` and `ValidationConfig`.

---

Let me know your thoughts.

**Comment by skim0119** [comment]
Could be considered for saving checkpoints:
https://github.com/google/orbax
